# 1장 2강: 핵심 모델과 선택 기준 압축 정리

## 실습 목표

- Ames Housing의 `SalePrice` 예측 상황에서 **Linear Regression**이 어떤 역할을 하는지 확인합니다.
- KNN과 K-means가 **거리 기반**이라는 점과 스케일에 민감하다는 특징을 구분합니다.
- Decision Tree의 규칙 기반 분할 특징과 스케일링이 필요 없다는 점을 확인합니다.
- K-means와 PCA의 목적이 각각 **군집화 / 차원축소**임을 구분합니다.
- 데이터의 목적과 특성에 따라 다섯 모델 중 **1차 후보를 선택하는 기준**을 정리합니다.

> 이번 실습은 1장 2강 교안에서 다룬 모델의 **핵심 성격과 선택 기준**만 사용합니다.  
> 교안에서 다루지 않은 성능지표, 교차검증, 하이퍼파라미터 탐색, 앙상블 모델은 사용하지 않습니다.


In [3]:
# 선형 회귀 : 입력 특성의 선형 결합, 절편으로 수치형 목표값을 예측하는 회귀모델
# → 공부 시간에 따른 1,2,3시간에 따른 성적 향상이 3,5,7점인 것을 생각

# 로지스틱 회귀 : 입력의 선형 결합을 바탕으로 클래스에 속할 확률을 추정하는 분류 모델
# → 고객 A의 구매 확률을 0.82로, 고객 B의 구매확률을 0.37로 예측했다면 Baseline(0.5)에 A는 구매, B는 미구매로 예측

# sigmoid : 실수 입력값을 0과 1사이의 값으로 바꾸는 함수
# → 이진 분류 로지스틱 회귀에서 선형 결합의 점수를 확률로 표현하는데 사용

# 비선형 관계 : 선형 관계로 표현할 수 없는 경우
# → 공부 시간이 1,2,3,4,5,6 시간에 따른 성적 향상이 10,20,30,40,50,60점이었는데 7시간부터 상승폭이 0이 되더니 8시간 부터는 -로 전환됨

# 결정트리(decision tree) : 특성(X)에 대한 조건으로 반복하여 나누어 예측(분류, 회귀)하는 모델
# → 방문 횟수가 1회 이상? → yes / no, 구매 금액이 0원 초과? → yes / no ....

# 노드와 리프 : 노드는 나무의 지점, 루트노드는 시작점, 리프노드는 최종 예측 지점을 의미
# → 질문을 시작하는 곳, 다시 질문하는 곳, 답을 내리는 곳

# 지니 불순도 : 한 노드 안에 클래스가 얼마나 섞여있는지 나타내는 값
# → 구매자는 구매자끼리, 미구매자는 미구매자들끼리 묶여야 지니 불순도가 낮아질 수 있음

# 과적합 : 학습데이터의 세부 특징이나 우연한 패턴까지 지나치게 맞춰 새 데이터에 대한 성능이 떨어지는 현상
# → 과적합인지 아닌지 학습 곡선으로 확인해볼 수 있음
# → 학습곡선 그래프를 그려서 val과 train데이터가 간격이 점점 더 멀어지는 경우를 과적합 상태로 의심
# → epoch당 loss를 봐서 loss가 줄어들지 않고 늘어나는 경우

# 앙상블 모델 : 여러 모델의 예측을 결합하여 최종 예측을 만드는 경우
# → 부트스트랩(tree), 배깅(여러 모델 결과를 결합)

# KNN(최근접 이웃) : 새 데이터와 가까운 K개 학습 사례의 정답을 이용해 예측하는 지도학습 모델 (주소, 좌표 찾기 등)
# → A 건물 내 이동거리가 5M이내인 좌표들을 A 건물 내 이동거리로 묶을 수 있다.
# → 새 고객과 가까운 구매, 구매, 미구매의 사람들이 있다면 동일한 가중치의 다수결로 구매를 예측한다.
# n_neighbors(이웃 수) : KNN이 예측할 때 참고하는 가장 가까운 학습 사례 수
# → A 건물 내 한 번의 움직임을 보였지만 GNSS 데이터에서 30회의 노드가 찍힌 경우, n_neighbors가 30으로 볼 수 있다. 
# → 가장 가까운 고객은 미구매지만 가까운 5명의 고객 중 4명이 구매라면 구매로 예측할 수 있다.

# 유클리드 거리 : 여러 특성으로 표현한 두 점 사이의 직선 거리
# → 좌표 계산을 통해 직선 사이의 거리를 구함
# → 새 고객의 방문 및 구매 횟수가 (8,3), (9,3) B가 (4,1)이라면 거리는 각각 1과 4.5임
# → A가 더 가까운 이웃으로 채택

# K-Means(군집화) : 정답없이(비지도학습) K개의 중심적으로 나누는 학습 방법
# → 연령별, 특성별로 군집화 (정답X)
# centroid(중심점) : 군집에 배정된 데이터의 평균 위치
# → 각 특성으로 평균 좌표를 만든다.

# 엘보(Elbow) : 군집화를 진행할 때 기울기가 급감하는 지점을 참고해 K의 후보를 고르는 방법

# PCA(주성분 분석) : 기존 특성(X)의 선형 조합으로 분산을 잘 나타내는 새 축을 만드는 방법
# → 유사한 특성들을 묶어서 하나의 컬럼으로 만들 때 사용
# → 고객 행동 특성 20개를 5개의 주성분으로 바꿔서 5개의 수치를 변수로 사용할 수 있다.


## 실습 준비

이번 실습에서도 **Ames Housing** 데이터를 사용합니다.

### 사용할 주요 컬럼

- `GrLivArea`: 지상 생활 면적
- `OverallQual`: 주택의 전반적인 품질
- `YearBuilt`: 건축 연도
- `LotArea`: 대지 면적
- `SalePrice`: 주택 판매 가격

먼저 데이터를 불러오고, 사용할 컬럼의 자료형·결측치·기초 통계량을 확인합니다.


In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

df = pd.read_csv("../Ames_Housing.csv")

use_cols = ["GrLivArea", "OverallQual", "YearBuilt", "LotArea", "SalePrice"]
data = df[use_cols].copy()

print("데이터 크기:", data.shape)
display(data.head())


데이터 크기: (1460, 5)


,GrLivArea,OverallQual,YearBuilt,LotArea,SalePrice
0,1710,7,2003,8450,208500
1,1262,6,1976,9600,181500
2,1786,7,2001,11250,223500
3,1717,7,1915,9550,140000
4,2198,8,2000,14260,250000


In [3]:
print("[자료형]")
display(data.dtypes.to_frame("dtype"))

print("[결측치 수]")
display(data.isna().sum().to_frame("missing_count"))

print("[기초 통계량]")
display(data.describe())


[자료형]


,dtype
GrLivArea,int64
OverallQual,int64
YearBuilt,int64
LotArea,int64
SalePrice,int64


[결측치 수]


,missing_count
GrLivArea,0
OverallQual,0
YearBuilt,0
LotArea,0
SalePrice,0


[기초 통계량]


,GrLivArea,OverallQual,YearBuilt,LotArea,SalePrice
count,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000
mean,1515.463699,6.099315,1971.267808,10516.828082,180921.195890
std,525.480383,1.382997,30.202904,9981.264932,79442.502883
min,334.000000,1.000000,1872.000000,1300.000000,34900.000000
25%,1129.500000,5.000000,1954.000000,7553.500000,129975.000000
50%,1464.000000,6.000000,1973.000000,9478.500000,163000.000000
75%,1776.750000,7.000000,2000.000000,11601.500000,214000.000000
max,5642.000000,10.000000,2010.000000,215245.000000,755000.000000


---

# 필수 1. Linear Regression으로 기준 모델의 성격 확인

## 배경

교안에서 **Linear Regression**은 연속적인 숫자를 예측하는 회귀 모델이며, 각 특성에 가중치를 곱해 더하는 구조라고 배웠습니다.  
또한 가중치를 확인할 수 있어 **해석이 쉬운 모델**이라는 특징이 있습니다.

Ames Housing에서 `GrLivArea`, `OverallQual`, `YearBuilt`를 이용해 `SalePrice`를 예측하는 Linear Regression을 만들어봅니다.

## 문제

### 요구사항

1. 입력 데이터 `X`는 `GrLivArea`, `OverallQual`, `YearBuilt`로 구성하고, 정답 `y`는 `SalePrice`로 지정하세요.
2. `train_test_split()`을 이용해 train과 test로 나누세요. `test_size=0.2`, `random_state=42`를 사용합니다.
3. `LinearRegression()` 모델을 만들고 train 데이터로 학습하세요.
4. 학습한 모델의 `intercept_`와 각 특성의 `coef_`를 출력하세요.
5. test 데이터의 앞 5개 행에 대해 예측값을 만들고 실제 `SalePrice`와 나란히 확인하세요.

### 확인 포인트

- `SalePrice`가 연속적인 숫자이므로 Linear Regression을 적용할 수 있는지 확인합니다.
- 계수(`coef_`)를 통해 각 입력 특성이 예측식에서 어떤 방향의 가중치를 갖는지 확인합니다.
- 이번 문제에서는 **모델 성능지표를 계산하지 않습니다.**

### Q&A

- **Q1.** Linear Regression은 회귀와 분류 중 어느 문제에 사용하는 모델인가요?
- **Q2.** Linear Regression이 해석이 쉬운 모델로 설명되는 이유는 무엇인가요?
- **Q3.** `SalePrice` 예측에 Logistic Regression을 사용하지 않는 이유는 무엇인가요?


In [4]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# TODO 1. X와 y를 만드세요.
X = data[["GrLivArea", "OverallQual", "YearBuilt"]]
y = data["SalePrice"]

print("X 크기:", X.shape, "| y 크기:", y.shape)

# TODO 2. train / test로 분리하세요.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
print(f"train: {len(X_train)}행 / test: {len(X_test)}행")


# TODO 3. Linear Regression을 학습하세요.
model = LinearRegression()
model.fit(X_train, y_train)


# TODO 4. 절편과 계수를 출력하세요.
print(f"\n절편(intercept_): {model.intercept_:,.2f}")

coef_table = pd.DataFrame({
    "feature": X.columns,
    "coef": model.coef_.round(2),
})
display(coef_table)

# TODO 5. test 앞 5개 행의 실제값과 예측값을 비교하세요.
y_pred_5 = model.predict(X_test.head())

comparison = pd.DataFrame({
    "실제 SalePrice": y_test.head().values,
    "예측 SalePrice": y_pred_5.round(0).astype(int),
    "차이(실제-예측)": (y_test.head().values - y_pred_5).round(0).astype(int),
}, index=X_test.head().index)
display(comparison)


X 크기: (1460, 3) | y 크기: (1460,)
train: 1168행 / test: 292행

절편(intercept_): -1,075,160.06


,feature,coef
0,GrLivArea,60.13
1,OverallQual,24520.16
2,YearBuilt,514.80


,실제 SalePrice,예측 SalePrice,차이(실제-예측)
892,154500,146735,7765
1105,325000,305180,19820
413,115000,101277,13723
522,159000,174337,-15337
1036,315500,276140,39360


### 필수 1 답변 작성

### 1번 표 해석
- GrLiveArea의 계수 60.13은 품질 및 건축연도가 같을 때 면적 1제곱피트 증가 당 예측 가격이 약 60.13달러가 증가
- OverallQual의 계수 24,520은 면적 및 건축연도가 같을 때 품질 점수 1점 증가 당 예측 가격이 약 24,520달러가 증가
- YearBuilt의 계수 514.7은 면적 및 품질이 같을 때 건축 연도가 1년 뒤일 수록 예측 가격이 약 514.7달러 증가
- OverallQual이 가장 큰 숫자이지만 단위 수가 다르기 때문에 가장 중요한 변수라고 결론을 내리면 안된다.

### 2번 표 해석
- 원본 행 인덱스 892: 실제 가격 154,500달러 예측 가격 146,734, 예측 - 실제 = -7,765 달러: 실제보다 약 7,765달러 낮게 예측
- 원본 행 인덱스 1105: 실제 가격 325,000달러 예측 가격 305,179, 예측 - 실제 = -19,820달러: 실제보다 약 19,820달러 낮게 예측

**Q1. Linear Regression의 문제 유형:**  
-> 회귀유형  

**Q2. 해석이 쉬운 이유:**  
-> 각 특성에 적용된 가중치(계수)를 직접적으로 확인이 가능하다. 
 
**Q3. Logistic Regression을 사용하지 않는 이유:**  
-> 연속적인 수치 범위를 직접 출력할 수 없기 때문


---

# 필수 2. K-means와 PCA로 비지도학습 목적 구분

## 배경

교안에서 **K-means**는 정답 레이블 없이 비슷한 데이터를 K개의 군집으로 묶는 군집화 알고리즘이라고 배웠습니다.  
거리로 가까운 중심을 찾기 때문에 **특성 스케일에 민감**하여 스케일링을 적용하는 것이 좋습니다.

반면 **PCA**는 많은 특성을 더 적은 수의 새로운 축으로 압축하는 **차원축소** 기법입니다.

같은 Ames Housing의 숫자형 특성을 이용해 두 기법이 결과를 어떻게 다르게 만드는지 확인합니다.

## 문제

### 요구사항

1. `GrLivArea`, `OverallQual`, `YearBuilt`, `LotArea` 네 컬럼으로 `X_unsup`을 만드세요. `SalePrice`는 사용하지 않습니다.
2. `StandardScaler()`로 `X_unsup`의 스케일을 맞추세요.
3. 스케일링한 데이터에 `KMeans(n_clusters=3, random_state=42, n_init=10)`을 적용하고 각 행의 군집 번호를 구하세요.
4. 군집별 데이터 개수를 출력하세요.
5. 원본 네 특성에 `PCA(n_components=2)`를 적용해 2개의 새로운 축으로 변환하세요.
6. PCA 적용 전과 적용 후 데이터의 크기(shape)를 출력하세요.

### 확인 포인트

- K-means에서는 정답 `y` 없이 특성 데이터만 사용합니다.
- K-means의 K는 **군집 개수**를 의미합니다.
- PCA는 데이터를 그룹으로 나누는 것이 아니라 **특성의 개수를 줄이는 것**이 목적입니다.
- 이번 실습에서 K-means의 K를 고르기 위한 Elbow/Silhouette 비교는 진행하지 않습니다.

### Q&A

- **Q1.** K-means는 지도학습과 비지도학습 중 어디에 해당하나요?
- **Q2.** K-means 전에 스케일링을 적용한 이유는 무엇인가요?
- **Q3.** K-means의 K와 KNN의 K는 각각 무엇의 개수인가요?
- **Q4.** PCA를 적용한 뒤 `(행 개수, 특성 개수)`에서 어떤 부분이 바뀌나요?


In [5]:
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA

# TODO 1. 비지도학습에 사용할 X_unsup을 만드세요.
X_unsup = data[["GrLivArea", "OverallQual", "YearBuilt", "LotArea"]]
print("X_unsup 크기:", X_unsup.shape)

# TODO 2. StandardScaler로 스케일을 맞추세요.
scaler = StandardScaler()
X_unsup_scaled = scaler.fit_transform(X_unsup)

# TODO 3. K-means로 3개 군집을 만들고 군집 번호를 구하세요.
kmeans = KMeans(n_clusters=3, random_state=42, n_init=10)
cluster_labels = kmeans.fit_predict(X_unsup_scaled)

# TODO 4. 군집별 데이터 개수를 출력하세요.
print("\n군집별 데이터 개수:")
print(pd.Series(cluster_labels).value_counts().sort_index())

# TODO 5. PCA로 4개 특성을 2개의 새로운 축으로 변환하세요.
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_unsup)

# TODO 6. PCA 적용 전/후 shape를 출력하세요.
print(f"\nPCA 적용 전 shape: {X_unsup.shape}")
print(f"PCA 적용 후 shape: {X_pca.shape}")

X_unsup 크기: (1460, 4)

군집별 데이터 개수:
0    819
1    637
2      4
Name: count, dtype: int64

PCA 적용 전 shape: (1460, 4)
PCA 적용 후 shape: (1460, 2)


### 필수 2 답변 작성

군집별 주택 수  
- 군집번호(0) : 주택 수 819개, 가장 큰 군집
- 군집번호(1) : 주택 수 637개, 두번째로 큰 군집
- 군집번호(2) : 주택 수 4개, 가장 적은 군집

-> 군집 크기가 다른 두 집단에 비해 굉장히 작은 4는 오류라고 봐야하는가? 아니면 잘못된 결과라고 해석해야하는가?  
-> 전체 0.3% 정도의 분포를 보이는 군집번호 2는 이상치나 잘못된 부분은 아니다.  

PCA (1460,4) -> (1460,2)  
- 관측대상 : 기존 (주택1,460채), PCA 후 (주택 1,460채)  
- 한 주택을 표현하는 특성X : 기존(4개), PCA 후(2개)  
- 열의 의미 : 기존(면적, 품질, 건축연도, 대지면적), PCA 후(원래 특성들을 조합한 새로운 축)  

Y를 설명하려던 변수 4개가 2개로 줄었는데 정보 손실이 발생한 것 아니냐?  
- PCA는 변수 4개에 대한 유사한 변동성을 가진 데이터를 묶으려는 과정  
- 4개의 변수를 2개로 줄이면 일부 변동성을 충분히 반영하지 못할 수 있지만  
- 이 부분이 반드시 정보의 50%를 손실했다고 볼 수 없습니다.


**Q1. K-means의 학습 방식:**  
- 비지도 학습  

**Q2. K-means 전에 스케일링이 필요한 이유:**  
- 원본 통계량에서 품질은 1~10점, 대지면적은 1,300~215,000정도로  
원본값에서 좌표 거리를 계산하면 숫자 변동이 매우 커지는 경우가 있다.  
표준화는 이러한 변동 규모와 단위를 줄여 영향 차이를 줄일 수 있다.  
단, 표준화가 이상치를 제거하거나 중요도를 동일시하게 만든다는 의미는 아니다.  
-> 표준화를해서 이상치를 제거한다는 말은 절대 아님

**Q3. K-means의 K / KNN의 K 차이:**  
- K-Means는 군집의 개수 / KNN은 이웃의 개수  

**Q4. PCA 적용 전후 shape 변화:**  
- 행 개수는 유지, Y를 설명하는 X변수가 4개->2개로 줄었다.


---

# 과제 1. Ames Housing 상황에 맞는 모델 후보 선택

## 배경

이번 강의의 핵심은 모델을 복잡하게 튜닝하는 것이 아니라, **데이터의 목적과 특성을 보고 어떤 모델부터 후보로 볼지 판단하는 것**입니다.

아래 상황은 모두 이번 강의에서 학습한 Linear Regression, Logistic Regression, KNN, Decision Tree, K-means, PCA의 특징만으로 판단할 수 있습니다.

## 문제

각 상황에서 가장 먼저 고려할 모델 또는 기법을 하나 선택하고, **교안에서 배운 선택 기준을 근거로 이유를 한 문장으로 작성하세요.**

### 요구사항

1. `SalePrice`처럼 연속적인 숫자를 예측하면서, 각 특성의 가중치를 확인해 예측 근거도 쉽게 설명하고 싶습니다. 어떤 모델을 먼저 고려해야 하나요?
2. 주택 특성 사이의 관계가 단순한 직선 형태가 아니며, 사람이 이해할 수 있는 `예/아니오` 규칙 형태로 나누고 싶습니다. 어떤 모델을 먼저 고려해야 하나요?
3. 거리 기반으로 주변의 비슷한 주택을 참고하여 예측하려고 합니다. 어떤 모델을 고려할 수 있으며, 사용 전에 무엇을 확인해야 하나요?
4. `SalePrice` 같은 정답 레이블을 사용하지 않고 주택을 비슷한 특성끼리 3개 그룹으로 묶고 싶습니다. 어떤 기법을 사용해야 하나요?
5. 숫자형 특성이 많아 분석하기 어렵기 때문에 핵심 정보를 유지하면서 2개의 새로운 축으로 줄이고 싶습니다. 어떤 기법을 사용해야 하나요?

### 확인 포인트

- 모델 이름만 쓰지 않고 **왜 그 모델을 선택했는지** 교안의 특성과 연결합니다.
- 성능지표 비교, 교차검증, GridSearch 등의 방법은 사용하지 않습니다.
- 과제는 필수 문제에서 학습한 **모델의 성격과 선택 기준을 다시 적용하는 수준**입니다.


### 과제 1 답변 작성

1. **연속값 예측 + 해석 중요:**  
- Linear Regression  
SalePrice가 연속적인 숫자 값이기 때문에 회귀 모델이 필요하고,  
각 특성에 곱해지는 계수(coef_)를 직접 확인할 수 있어 "이 x가 1 늘어나면 예측 가격이 얼마만큼 변한다"로 근거를 설명할 수 있기 때문  

2. **비선형 관계 + 규칙 해석:**  
- Decision Tree  
"생활 면적이 1500 이상인가?" 같은 예/아니오 질문을 반복해 데이터를 나누는 구조라 직선으로 표현되지 않는 관계도 잡아낼 수 있고, 그 분기 규칙을 사람이 그대로 읽고 이해할 수 있기 때문  

3. **가까운 이웃 기반 예측:**  
- KNN  
새 데이터와 가장 가까운 K개 이웃을 찾아 그 값들로 예측하는 모델, 거리로 판단하기 때문에 특성마다 단위와 범위가 다르면 범위가 큰 특성이 거리를 독차지하므로 사용 전에 StandardScaler 등으로 스케일을 맞췄는지 반드시 확인  

4. **정답 없이 3개 그룹으로 군집화:**  
- K-means  
정답 레이블(SalePrice)없이 특성 데이터만으로 비슷한 것끼리 묶는 비지도 군집화, n_clusters=3으로 원하는 군집 개수를 직접 지정할 수 있음  

5. **많은 특성을 2개 축으로 압축:**  
- PCA  
분산이 가장 큰 방향을 새로운 축으로 삼아 핵심 정보를 최대한 유지하면서 특성 개수를 줄이는 차원축소 기법, n_components=2로 2개 축까지 압축할 수 있기 때문

---

# 실습 마무리

아래 내용을 한 문장씩 정리해보세요.

1. **Linear Regression은 어떤 상황에서 1차 후보가 될 수 있는가?**  
- 회귀 문제, 각 특성의 가중치를 통해 결과를 쉽게 설명하고 빠르게 결과를 받을 때 1차 후보로 고려  

2. **KNN / Decision Tree는 데이터 특성에 따라 어떤 차이가 있는가?**  
- KNN : 거리 기반이므로 스케일링에 민감,  
- Decision Tree : 규칙 기반으로 데이터를 나누기 때문에 스케일링이 필요하지 않음

3. **K-means와 PCA의 목적은 어떻게 다른가?**  
- K-Means : 데이터를 비슷한 것끼리 군집으로 묶는 것  
- PCA : 많은 특성(X)를 더 적은 새로운 축으로 줄이는 것  

4. **처음부터 복잡한 모델을 선택하기보다 후보를 좁힐 때 어떤 기준을 볼 수 있는가?**  
- 문제의 목적(회귀/분류)확인, 비선형성인지 선형성인지 확인, 스케일링 필요한지, 차원축소가 필요한지